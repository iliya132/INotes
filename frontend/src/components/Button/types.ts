import { MouseEventHandler } from "react";

export interface IButtonProps {
    title: string,
    className?: string,
    disabled?: boolean;
    onClick?: MouseEventHandler<HTMLButtonElement>;
}