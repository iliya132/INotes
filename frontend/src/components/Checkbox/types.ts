/* eslint-disable no-unused-vars */
export interface ICheckboxProps {
    defaultValue?: boolean;
    id: string;
    className?: string;
    onChanged? : (newValue:boolean) => void;
    title?: string;
}
