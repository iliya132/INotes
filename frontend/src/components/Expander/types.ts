export interface IExpanderProps {
    isExpanded: boolean;
    expanderBody: JSX.Element;
    children: JSX.Element;
    onClick?: () => void;
}
