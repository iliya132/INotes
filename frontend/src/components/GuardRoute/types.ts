export interface IGuardRouteProps {
    canActivate: boolean;
    redirectTo: string;
    children: JSX.Element;
    onSuccess?: () => void
  }
  