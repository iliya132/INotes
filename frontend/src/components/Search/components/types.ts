export interface IFoundResultProps {
    title: string;
    context: string;
    onClick: () => void;
}
