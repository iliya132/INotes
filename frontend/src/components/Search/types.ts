import React from "react";

export default interface ISearchProps {
    onChange?: React.ChangeEventHandler<HTMLInputElement>,
    ref?: any

}