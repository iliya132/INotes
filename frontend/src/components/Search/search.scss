@import '../../Misc/colors.scss';

.search-container {
    display: grid;
    grid-template-columns: 1fr 50px;
}

.search-field {
    width: 100%;
    height: 100%;
    padding-left: 10px;
    border: none;
    outline: none;
    background-color: transparent;
    color: $TextColor;
    font-size: 13px;
    font-size: 12px;
    line-height: 13px;
}

.input-container {
    display: grid;
    grid-template-columns: 1fr 50px;
    width: 98%;
    max-width: 400px;
    height: 100%;
    border-radius: 6px;
    background-color: rgb(218 215 205 / 80%) !important;
}

.eye-icon {
    position: relative;
    grid-column: 2;
    width: 48px;
    height: 48px;
    margin-top: auto;
    cursor: pointer;
}
