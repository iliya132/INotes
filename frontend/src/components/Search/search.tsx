import React, { useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import Popup from 'reactjs-popup';
import notesController from '../../controllers/NotesController';
import { ISearchResult } from '../../controllers/types';
import { allNotes } from '../../store/reducers/notebooksReducer';
import Svg from '../Svg';
import { Icons } from '../Svg/types';
import FoundResult from './components';
import styles from './search.scss';
import ISearchProps from './types';

export function Search(props: ISearchProps) {
    const { onChange } = props;
    const inputRef = useRef(null);
    const nav = useNavigate();
    const allNotesState = useSelector(allNotes);
    const [foundResults, setFoundResults] = useState({
        notebooks: [],
        notes: [],
    } as ISearchResult);
    const [isOpen, setOpen] = useState(false);
    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        
        const found = notesController.find(e.target.value)
        setFoundResults(found);
        if (onChange) {
            onChange(e);
        }
        setOpen(true);
    };

    const handleClose = () => {
        setOpen(false);
    };

    const clearSearchInput = () => {
        (inputRef!.current! as HTMLInputElement).value = '';
        setFoundResults({ notebooks: [], notes: [] });
    };

    const handleSelectNote = (noteId: number) => {
        nav("/" + noteId);
    }

    const handleSelectNotebook = (notebookId: number) => {
        const firstNote = allNotesState.filter(it=> it.parent.id === notebookId)[0];
        handleSelectNote(firstNote.id);
    }

    const isAnyResultsFound = foundResults.notebooks.length > 0 || foundResults.notes.length > 0;

    return (
        <Popup
            onClose={handleClose}
            open={isOpen}
            position="bottom left"
            trigger={
                <div className={styles['input-container']} data-testid="search-container">
                    <input
                        ref={inputRef}
                        id="search"
                        type="text"
                        className={styles['search-field']}
                        placeholder="Type here to search..."
                        onChange={(e) => {
                            handleInputChange(e);
                        }}
                        autoComplete="disabled"
                        data-testid="search-input"
                    />
                    <Svg icon={Icons.Search} className={styles['eye-icon']} />
                </div>
            }
            arrow={false}>
            {(close) => {
                return !isAnyResultsFound ? undefined : (
                    <div className={styles['popup-container']}>
                        {foundResults.notebooks.map((notebook) => (
                            <FoundResult
                                title={notebook.notebook.name}
                                context={notebook.context}
                                key={`notebook_result_${notebook.notebook.id}`}
                                onClick={() => {
                                    close();
                                    clearSearchInput();
                                    handleSelectNotebook(notebook.notebook.id);
                                }}
                            />
                        ))}
                        {foundResults.notes.map((note) => (
                            <FoundResult
                                title={note.note.name}
                                context={note.context}
                                key={`note_result_${note.note.id}`}
                                onClick={() => {
                                    close();
                                    clearSearchInput();
                                    handleSelectNote(note.note.id);
                                }}
                            />
                        ))}
                    </div>
                );
            }}
        </Popup>
    );
}
