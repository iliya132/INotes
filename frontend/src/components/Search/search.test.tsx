/* eslint-disable no-undef */
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { RootState } from '../../store/store';
import { INote, INotebook } from '../../store/types';
import { Search } from './search';
import { authSlice } from '../../store/reducers/authReducer';
import { notesSlice } from '../../store/reducers/notebooksReducer';
import { configureStore } from '@reduxjs/toolkit';
import { MemoryRouter, Route, Routes, useParams } from 'react-router-dom';
import { IFoundNote, IFoundNotebook, ISearchResult } from '../../controllers/types';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import notesController from '../../controllers/NotesController';

//#region common setup
jest.mock('../../controllers/NotesController', () => ({ find: jest.fn() }));

let initialState: RootState = {
    authReducer: {
        error: '',
        isAuth: true,
        user: { roles: ['user'], userName: 'test-user', avatarUrl: '/default.png' },
        validation: {
            isSucceded: true,
            errors: {},
        },
    },
    notebooksReducer: {
        notebooks: [],
        allNotes: [],
    },
};

let store = configureStore({
    reducer: {
        authReducer: authSlice(initialState.authReducer).reducer,
        notebooksReducer: notesSlice(initialState.notebooksReducer).reducer,
    },
});
beforeEach(() => {
    store = configureStore({
        reducer: {
            authReducer: authSlice(initialState.authReducer).reducer,
            notebooksReducer: notesSlice(initialState.notebooksReducer).reducer,
        },
    });
});
//#endregion

describe('search field tests', () => {
    it('can be rendered', () => {
        const currentState = getSomeNotes(initialState);
        store = configureStore({
            reducer: {
                authReducer: authSlice(initialState.authReducer).reducer,
                notebooksReducer: notesSlice(currentState.notebooksReducer).reducer,
            },
        });
        act(() => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Routes>
                            <Route path="/" element={<Search />} />
                        </Routes>
                    </MemoryRouter>
                </Provider>,
            );
        });
    });

    it('should match snapshot', () => {
        const currentState = getSomeNotes(initialState);
        store = configureStore({
            reducer: {
                authReducer: authSlice(initialState.authReducer).reducer,
                notebooksReducer: notesSlice(currentState.notebooksReducer).reducer,
            },
        });
        act(() => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Routes>
                            <Route path="/" element={<Search />} />
                        </Routes>
                    </MemoryRouter>
                </Provider>,
            );
        });
        expect(document.body.innerHTML).toMatchSnapshot('search field snapshot');
    });

    it('can find notebook', async () => {
        const find = jest.fn((text: string): ISearchResult => {
            const noteBookState = store.getState().notebooksReducer;
            const notes = noteBookState.allNotes;
            const notebooks = noteBookState.notebooks;
            const foundNotes = notes
                .filter((it) => filterNotesContains(it, text))
                .map((it) => getFoundNoteContext(it, text));
            const foundNotebooks = notebooks
                .filter((it) => filterNotebookContains(it, text))
                .map((it) => getFoundNotebookContext(it, text));
            return {
                notes: foundNotes,
                notebooks: foundNotebooks,
            };
        });
        notesController.find = find;

        const user = userEvent.setup();
        const currentState = getSomeNotes(initialState);
        store = configureStore({
            reducer: {
                authReducer: authSlice(initialState.authReducer).reducer,
                notebooksReducer: notesSlice(currentState.notebooksReducer).reducer,
            },
        });
        act(() => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Routes>
                            <Route path="/" element={<Search />} />
                        </Routes>
                    </MemoryRouter>
                </Provider>,
            );
        });
        const searchContainer = screen.getByTestId('search-container');
        const input = screen.getByTestId('search-input');
        expect(searchContainer).not.toBeNull();
        expect(input).not.toBeNull();

        let notebookToFind: string = currentState.notebooksReducer.notebooks[0].name;
        await act(async () => {
            await user.click(searchContainer);
            await user.type(input, notebookToFind);
        });
        const popupContainer = document.body!.querySelector('.popup-container');
        expect(popupContainer).not.toBeNull();
        expect(popupContainer?.textContent).toContain(notebookToFind);
    });

    it('can find note', async () => {
        const find = jest.fn((text: string): ISearchResult => {
            const noteBookState = store.getState().notebooksReducer;
            const notes = noteBookState.allNotes;
            const notebooks = noteBookState.notebooks;
            const foundNotes = notes
                .filter((it) => filterNotesContains(it, text))
                .map((it) => getFoundNoteContext(it, text));
            const foundNotebooks = notebooks
                .filter((it) => filterNotebookContains(it, text))
                .map((it) => getFoundNotebookContext(it, text));
            return {
                notes: foundNotes,
                notebooks: foundNotebooks,
            };
        });
        notesController.find = find;

        const user = userEvent.setup();
        const currentState = getSomeNotes(initialState);
        store = configureStore({
            reducer: {
                authReducer: authSlice(initialState.authReducer).reducer,
                notebooksReducer: notesSlice(currentState.notebooksReducer).reducer,
            },
        });
        act(() => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Routes>
                            <Route path="/" element={<Search />} />
                        </Routes>
                    </MemoryRouter>
                </Provider>,
            );
        });
        const searchContainer = screen.getByTestId('search-container');
        const input = screen.getByTestId('search-input');
        expect(searchContainer).not.toBeNull();
        expect(input).not.toBeNull();

        let noteToFind: string = currentState.notebooksReducer.allNotes[0].content.substring(0, 5);
        await act(async () => {
            await user.click(searchContainer);
            await user.type(input, noteToFind);
        });
        const popupContainer = document.body!.querySelector('.popup-container');
        expect(popupContainer).not.toBeNull();
        expect(popupContainer?.textContent).toContain(noteToFind);
    });

    it('should close on click outside', async () => {
        const find = jest.fn((text: string): ISearchResult => {
            const noteBookState = store.getState().notebooksReducer;
            const notes = noteBookState.allNotes;
            const notebooks = noteBookState.notebooks;
            const foundNotes = notes
                .filter((it) => filterNotesContains(it, text))
                .map((it) => getFoundNoteContext(it, text));
            const foundNotebooks = notebooks
                .filter((it) => filterNotebookContains(it, text))
                .map((it) => getFoundNotebookContext(it, text));
            return {
                notes: foundNotes,
                notebooks: foundNotebooks,
            };
        });
        notesController.find = find;

        const user = userEvent.setup();
        const currentState = getSomeNotes(initialState);
        store = configureStore({
            reducer: {
                authReducer: authSlice(initialState.authReducer).reducer,
                notebooksReducer: notesSlice(currentState.notebooksReducer).reducer,
            },
        });
        act(() => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Routes>
                            <Route path="/" element={<Search />} />
                        </Routes>
                    </MemoryRouter>
                </Provider>,
            );
        });
        const searchContainer = screen.getByTestId('search-container');
        const input = screen.getByTestId('search-input');
        expect(searchContainer).not.toBeNull();
        expect(input).not.toBeNull();

        let noteToFind: string = currentState.notebooksReducer.allNotes[0].content.substring(0, 5);
        await act(async () => {
            await user.click(searchContainer);
            await user.type(input, noteToFind);
        });
        let popupContainer = document.body!.querySelector('.popup-container');
        expect(popupContainer).not.toBeNull();
        expect(popupContainer?.textContent).toContain(noteToFind);

        await act(async () => {
            await user.click(document.body);
        });

        popupContainer = document.body!.querySelector('.popup-container');
        expect(popupContainer).toBeNull();
    });

    it('should clear found results on click', async () => {
        const find = jest.fn((text: string): ISearchResult => {
            const noteBookState = store.getState().notebooksReducer;
            const notes = noteBookState.allNotes;
            const notebooks = noteBookState.notebooks;
            const foundNotes = notes
                .filter((it) => filterNotesContains(it, text))
                .map((it) => getFoundNoteContext(it, text));
            const foundNotebooks = notebooks
                .filter((it) => filterNotebookContains(it, text))
                .map((it) => getFoundNotebookContext(it, text));
            return {
                notes: foundNotes,
                notebooks: foundNotebooks,
            };
        });
        notesController.find = find;

        const user = userEvent.setup();
        const currentState = getSomeNotes(initialState);
        store = configureStore({
            reducer: {
                authReducer: authSlice(initialState.authReducer).reducer,
                notebooksReducer: notesSlice(currentState.notebooksReducer).reducer,
            },
        });
        act(() => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Routes>
                            <Route path="/" element={<Search />} />
                            <Route
                                path="/:number"
                                element={<Print/>}
                            />
                        </Routes>
                    </MemoryRouter>
                </Provider>,
            );
        });
        const searchContainer = screen.getByTestId('search-container');
        const input = screen.getByTestId('search-input');
        expect(searchContainer).not.toBeNull();
        expect(input).not.toBeNull();

        let noteToFind: string = currentState.notebooksReducer.allNotes[0].content.substring(0, 5);
        await act(async () => {
            await user.click(searchContainer);
            await user.type(input, noteToFind);
        });
        let resultContainer = document.body!.querySelector('.result-container');
        expect(resultContainer).not.toBeNull();
        expect(resultContainer?.textContent).toContain(noteToFind);

        await act(async () => {
            await user.click(resultContainer!);
        });

        let popupContainer = document.body!.querySelector('.popup-container');
        expect(popupContainer).toBeNull();

        await act(async () => {
            await user.click(searchContainer);
        });

        popupContainer = document.body!.querySelector('.popup-container');
        expect(popupContainer).toBeNull();
        expect(document.body.innerHTML).toContain(currentState.notebooksReducer.allNotes[0].id.toString())
    });
});

function getSomeNotes(state: RootState): RootState {
    const result: RootState = {
        authReducer: state.authReducer,
        notebooksReducer: randomNotebookWithNotes(),
    };
    return result;
}

function randomNotebookWithNotes(): { notebooks: INotebook[]; allNotes: INote[] } {
    const notebooks: INotebook[] = [];
    const notes: INote[] = [];
    for (let i = 0; i < 10; i++) {
        notebooks.push(randomNotebook());
    }

    for (let i = 0; i < notebooks.length; i++) {
        for (let j = 0; j < 5; j++) {
            const currentNotebook = notebooks[i];
            notes.push(randomNote(currentNotebook));
        }
    }

    return {
        notebooks: notebooks,
        allNotes: notes,
    };
}

function randomNotebook(): INotebook {
    return {
        color: 'black',
        id: Math.random() * 150,
        name: randomString(),
    };
}

function randomNote(parent: INotebook): INote {
    return {
        content: randomString(),
        id: Math.random() * 100,
        isPublicUrlShared: false,
        isNew: false,
        parent: parent,
        name: randomString(),
        publicUrl: '',
    };
}

function randomString(): string {
    return (Math.random() + 1).toString(36);
}

function filterNotesContains(note: INote, text: string): boolean {
    return note.content.toLowerCase().includes(text.toLowerCase());
}

function getFoundNoteContext(note: INote, text: string): IFoundNote {
    return {
        note,
        context: getContext(note.content, text),
    };
}

function getFoundNotebookContext(notebook: INotebook, text: string): IFoundNotebook {
    return {
        notebook,
        context: getContext(notebook.name, text),
    };
}

const searchContextOffset = 5;
const minSearchContextOffset = 100;

function getContext(content: string, search: string): string {
    const foundIndex = content.toLowerCase().indexOf(search);
    const endIndex = foundIndex + search.length;
    let indexOfContext = foundIndex;
    let indexOfEndContext = endIndex + searchContextOffset;

    if (search.length < minSearchContextOffset) {
        const diff = minSearchContextOffset - search.length;

        indexOfContext -= diff / 2;
        indexOfEndContext += diff / 2;
        if (indexOfContext < 0) {
            indexOfEndContext -= indexOfContext;
        }
        if (indexOfEndContext > content.length) {
            indexOfContext -= indexOfEndContext - content.length;
        }
    }

    if (indexOfContext < 0) {
        indexOfContext = 0;
    }

    if (indexOfEndContext > content.length) {
        indexOfEndContext = content.length;
    }
    return (
        content.substring(indexOfContext, foundIndex) +
        '%%' +
        content.substring(foundIndex, endIndex) +
        '%%' +
        content.substring(endIndex, indexOfEndContext)
    );
}

function filterNotebookContains(notebook: INotebook, text: string): boolean {
    return notebook.name.toLowerCase().includes(text.toLowerCase());
}

function Print() {
    var fromPath = useParams();
    return <div>{fromPath.number}</div>;
}