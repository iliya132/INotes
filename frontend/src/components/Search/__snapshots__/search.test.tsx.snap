// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search field tests should match snapshot: search field snapshot 1`] = `"<div><div class=\\"input-container\\" data-testid=\\"search-container\\" aria-describedby=\\"popup-2\\"><input id=\\"search\\" type=\\"text\\" class=\\"search-field\\" placeholder=\\"Type here to search...\\" autocomplete=\\"disabled\\" data-testid=\\"search-input\\"><svg class=\\"image eye-icon tooltip\\" fill=\\"BlackCommon\\" width=\\"48\\" height=\\"48\\"><use xlink:href=\\"test-file-stub#search\\"></use></svg></div></div>"`;
